libvast/src/segment.cpp

```diff
@@ -0,0 +1,201 @@
+/******************************************************************************
+ *                    _   _____   __________                                  *
+ *                   | | / / _ | / __/_  __/     Visibility                   *
+ *                   | |/ / __ |_\ \  / /          Across                     *
+ *                   |___/_/ |_/___/ /_/       Space and Time                 *
+ *                                                                            *
+ * This file is part of VAST. It is subject to the license terms in the       *
+ * LICENSE file found in the top-level directory of this distribution and at  *
+ * http://vast.io/license. No part of VAST, including this file, may be       *
+ * copied, modified, propagated, or distributed except according to the terms *
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#include <caf/stream_deserializer.hpp>
+
+#include <caf/detail/scope_guard.hpp>
+
+#include "vast/bitmap_algorithms.hpp"
+#include "vast/const_table_slice_handle.hpp"
+#include "vast/ids.hpp"
+#include "vast/segment.hpp"
+#include "vast/si_literals.hpp"
+#include "vast/table_slice.hpp"
+
+#include "vast/detail/assert.hpp"
+#include "vast/detail/byte_swap.hpp"
+#include "vast/detail/coded_deserializer.hpp"
+#include "vast/detail/coded_serializer.hpp"
+#include "vast/detail/narrow.hpp"
+
+namespace vast {
+
+using namespace binary_byte_literals;
+
+namespace {
+
+template <class T>
+T to_little_endian(T x) {
```

Comment:
Why introduce yet another endian conversion? There's already `to_host_order` and `to_network_order` in `byte_swap.hpp`.

Reply:
Neither of those can properly model the case at hand: convert from little to host endian and back.

Reply:
> from little to host endian

I don't understand what you are trying to tell me. Intel is little endian, so if you wish to write little endian to disk you need not do anything. However, I think we should write in network order, i.e., big endian, to disk for portability. The same reason why the network uses big endian.

Reply:
> Intel is little endian, so if you wish to write little endian to disk you need not do anything.

I don't want to assume a fixed endianess (or restrict us to Intel), just optimize for little endian.

> However, I think we should write in network order, i.e., big endian, to disk for portability. 

I disagree here. If you look at any modern serialization library, they write in little endian because it covers the majority of modern CPU architectures and will thus lead to less byte swapping.